File: 2022/23/rust/src/grove.rs
use crate::Point;
use std::collections::HashMap;
use std::collections::HashSet;
use std::fs::File;
use std::io::{ BufRead, BufReader };

// --- Constants ---
const OFFSET: usize = 10_000;

pub struct Grove {
    elves: HashSet<Point>,
    proposed: HashMap<Point, Point>,
    banned: HashSet<Point>,
    round: usize
}

impl Grove {
    // --- Creates a Grove from an input file ---
    pub fn load(filename: &str) -> std::io::Result<Self> {
        let mut elves = HashSet::new();
        let proposed = HashMap::new();
        let banned = HashSet::new();
        let round = 0;

        let file = File::open(filename)?;
        for (y, line) in BufReader::new(file).lines().enumerate() {
            for (x, c) in line.unwrap().chars().enumerate() {
                if c == '#' {
                    let x = (OFFSET + x) as isize;
                    let y = (OFFSET + y) as isize;
                    elves.insert((x, y));
                }
            }
        }

        Ok(Grove{ elves, proposed, banned, round })
    }

    // --- Simulates one full round ---
    pub fn sim(&mut self) -> bool {
        let list: Vec<&Point> = self.elves.iter().collect();
        let mut result = false;

        // --- Phase 1: Considers move ---
        for pos in list {
            let mut found = false;
            for y in -1..=1 {
                for x in -1..=1 {
                    if x == 0 && y == 0 { continue; }
                    let p = (pos.0 + x, pos.1 + y);
                    if self.elves.contains(&p) {
                        found = true;
                        break;
                    }
                }
                if found { break; }
            }
            if !found {
                // --- No Elf adjacent, then do nothing ---
                self.proposed.insert(*pos, *pos);
                continue;
            }

            // --- Look at the 4 directions ---
            let mut p: Option<Point> = None;
            for n in 0..4 {
                let i = (n + self.round) % 4;
                if i == 0 { // North
                    if  self.elves.contains(&(pos.0 - 1, pos.1 - 1)) ||
                        self.elves.contains(&(pos.0    , pos.1 - 1)) ||
                        self.elves.contains(&(pos.0 + 1, pos.1 - 1)) { continue; }
                    p = Some((pos.0, pos.1 - 1));
                    break;
                } else if i == 1 { // South
                    if  self.elves.contains(&(pos.0 - 1, pos.1 + 1)) ||
                        self.elves.contains(&(pos.0    , pos.1 + 1)) ||
                        self.elves.contains(&(pos.0 + 1, pos.1 + 1)) { continue; }
                    p = Some((pos.0, pos.1 + 1));
                    break;
                } else if i == 2 { // West
                    if  self.elves.contains(&(pos.0 - 1, pos.1 - 1)) ||
                        self.elves.contains(&(pos.0 - 1, pos.1    )) ||
                        self.elves.contains(&(pos.0 - 1, pos.1 + 1)) { continue; }
                    p = Some((pos.0 - 1, pos.1));
                    break;
                } else if i == 3 { // East
                    if  self.elves.contains(&(pos.0 + 1, pos.1 - 1)) ||
                        self.elves.contains(&(pos.0 + 1, pos.1    )) ||
                        self.elves.contains(&(pos.0 + 1, pos.1 + 1)) { continue; }
                    p = Some((pos.0 + 1, pos.1));
                    break;
                } else {
                    unreachable!();
                }
            }

            if let Some(p) = p {
                if self.banned.contains(&p) {
                    self.proposed.insert(*pos, *pos);
                } else if self.proposed.contains_key(&p) {
                    self.banned.insert(p);
                    let e = self.proposed.remove(&p).unwrap();
                    self.proposed.insert(e, e);
                    self.proposed.insert(*pos, *pos);
                } else {
                    self.proposed.insert(p, *pos);
                }
                result = true;
            } else {
                self.proposed.insert(*pos, *pos);
            }
        }

        // --- Phase 2: Move ---
        self.elves.clear();
        for pos in self.proposed.keys() {
            self.elves.insert(*pos);
        }
        self.proposed.clear();
        self.banned.clear();
        self.round += 1;

        result
    }

    // --- Calculates the view rectangle ---
    fn find_view_rect(&self) -> (usize, usize, usize, usize) {
        let mut min_x = usize::max_value();
        let mut min_y = usize::max_value();
        let mut max_x = 0;
        let mut max_y = 0;
        for p in &self.elves {
            min_x = usize::min(min_x, p.0 as usize);
            min_y = usize::min(min_y, p.1 as usize);
            max_x = usize::max(max_x, p.0 as usize);
            max_y = usize::max(max_y, p.1 as usize);
        }
        (min_x, min_y, max_x, max_y)
    }

    // --- Count the empty titles ---
    pub fn empty_count(&self) -> usize {
        let mut total = 0;
        let (min_x, min_y, max_x, max_y) = self.find_view_rect();
        for y in min_y ..= max_y {
            for x in min_x ..= max_x {
                if !self.elves.contains(&(x as isize, y as isize)) { total += 1 };
            }
        }
        total
    }
}

File: 2022/23/javascript/grove.mjs
function Grove(text) {
	this.elves = new Set();
	this.proposed = {};
	this.banned = new Set();
	this.round = 0;

	for (const [y, line] of text.trim().split("\n").entries()) {
		for (const [x, c] of line.split("").entries()) {
			if (c === '#') this.elves.add(str(x, y));
		}
	}

	// --- Simulates one full round ---
	this.sim = function() {
		let result = false;

		// --- Phase 1: Considers move ---
		for (const position of this.elves) {
			const pos = point(position);
			let found = false;
			for (let y = -1; y <= 1; y++) {
				for (let x = -1; x <= 1; x++) {
					if (x === 0 && y === 0) continue;
					const p = str(pos.x, pos.y, x, y);
					if (this.elves.has(p)) {
						found = true;
						break;
					}
				}
				if (found) break;
			}
			if (!found) {
				// --- No Elf adjacent, then do nothing ---
				this.proposed[position] = position;
				continue;
			}

			// --- Look at the 4 directions ---
			let p = null;
			for (let n = 0; n < 4; n++) {
				let i = (n + this.round) % 4;
				if (i === 0) { // North
					if (this.elves.has(str(pos.x, pos.y, -1, -1)) ||
						this.elves.has(str(pos.x, pos.y,  0, -1)) ||
						this.elves.has(str(pos.x, pos.y,  1, -1))) continue;
					p = str(pos.x, pos.y, 0,  -1);
					break;
				} else if (i === 1) { // South
					if (this.elves.has(str(pos.x, pos.y, -1,  1)) ||
						this.elves.has(str(pos.x, pos.y,  0,  1)) ||
						this.elves.has(str(pos.x, pos.y,  1,  1))) continue;
					p = str(pos.x, pos.y, 0, 1);
					break;
				} else if (i === 2) { // West
					if (this.elves.has(str(pos.x, pos.y, -1, -1)) ||
						this.elves.has(str(pos.x, pos.y, -1,  0)) ||
						this.elves.has(str(pos.x, pos.y, -1,  1))) continue;
					p = str(pos.x, pos.y, -1, 0);
					break;
				} else if (i === 3) { // East
					if (this.elves.has(str(pos.x, pos.y,  1, -1)) ||
						this.elves.has(str(pos.x, pos.y,  1,  0)) ||
						this.elves.has(str(pos.x, pos.y,  1,  1))) continue;
					p = str(pos.x, pos.y, 1, 0);
					break;
				} else {
					throw "Unreachable!";
				}
			}

			if (p != null) {
				if (this.banned.has(p)) {
					this.proposed[position] = position;
				} else if (this.proposed.hasOwnProperty(p)) {
					this.banned.add(p);
					const e = this.proposed[p];
					delete this.proposed[p];
					this.proposed[e] = [e];
					this.proposed[position] = position;
				} else {
					this.proposed[p] = position;
				}
				result = true;
			} else {
				this.proposed[position] = position;
			}
		}

		// --- Phase 2: Move ---
		this.elves.clear();
		for (const position of Object.keys(this.proposed)) {
			this.elves.add(position);
		}
		this.proposed = {};
		this.banned.clear();
		this.round++;

		return result;
	};

	// --- Calculates the view rectangle ---
	this.findViewRect = function() {
		let minX = Number.MAX_VALUE;
		let minY = Number.MAX_VALUE;
		let maxX = 0;
		let maxY = 0;
		for (const pos of this.elves) {
			const p = point(pos);
			minX = Math.min(minX, p.x);
			minY = Math.min(minY, p.y);
			maxX = Math.max(maxX, p.x);
			maxY = Math.max(maxY, p.y);
		}
		return [ minX, minY, maxX, maxY ];
	};

	// --- Count the empty tiles --
	this.emptyCount = function() {
		let total = 0;
		let [ minX, minY, maxX, maxY ] = this.findViewRect();
		for (let y = minY; y <= maxY; y++) {
			for (let x = minX; x <= maxX; x++) {
				if (!this.elves.has(str(x, y))) total++;
			}
		}
		return total;
	};

	// --- Prints the grove map ---
	this.print = function() {
		let elves = 0;
		let [ minX, minY, maxX, maxY ] = this.findViewRect();
		for (let y = minY; y <= maxY; y++) {
			let line = "";
			for (let x = minX; x <= maxX; x++) {
				if (this.elves.has(str(x, y))) { elves++; line += "#"; }
				else line += ".";
			}
			console.log(line);
		}
		console.log("Elves: ", elves);
	}


}

// --- Helper functions ---
function str(x, y, dx = 0, dy = 0) {
	return (x + dx) + "," + (y + dy);
}

function point(str) {
	const i = str.indexOf(",");
	const x = Number(str.substring(0, i));
	const y = Number(str.substring(i + 1));
	return { x, y };
}

export default Grove;
